### PR TITLE
Fft small

### PR DESCRIPTION
**Pull request by albinahlback**
@fredrik-johansson @tthsqe12 

Edit: I changed C99 to C11 as C99 does not support `aligned_alloc`.

### CODE REVIEW

**Comment by albinahlback** [comment]
Looks like it is passing. Let me know what you think @fredrik-johansson.

Personally, I think this addition is OK, but it needs a big cleanup. I don't understand the code, and if Dan will not maintain this part of the code, someone else needs to clean this up.

File: src/nmod_poly/test/t-divrem_newton_n_preinv.c [hunk 1]
```diff
@@ -73,7 +73,7 @@ main(void)
         nmod_poly_clear(r);
         nmod_poly_clear(test);
     }
-
+#if 0
```
Comment (fredrik-johansson) [inline src/nmod_poly/test/t-divrem_newton_n_preinv.c:76]: I wonder why this was disabled?

**Comment by fredrik-johansson** [comment]
* There are some implicit declaration warnings in the fft_small tests.

* We'll need a ``FLINT_HAVE_FFT_SMALL`` define or such, for code where we want to call the fft_small functions if they are available

* Did you manage to enable Dan's optimization flags for the fft_small module?

**Comment by fredrik-johansson** [comment]
> Personally, I think this addition is OK, but it needs a big cleanup. I don't understand the code, and if Dan will not maintain this part of the code, someone else needs to clean this up.

The main problem is that documentation is scarce or cryptic (also an issue for parts of the mpoly code).

File: src/nmod_poly/test/t-divrem_newton_n_preinv.c (hunk 1 above)
Comment (albinahlback) [reply src/nmod_poly/test/t-divrem_newton_n_preinv.c:76]: Haha, now I do too...

**Comment by albinahlback** [comment]
> * There are some implicit declaration warnings in the fft_small tests.

Oh, I'll have to check that out.

> * We'll need a `FLINT_HAVE_FFT_SMALL` define or such, for code where we want to call the fft_small functions if they are available

This is easy.

> * Did you manage to enable Dan's optimization flags for the fft_small module?

So I think we definitely should optimize this module, but we need to create a profiler in order to see what works. `-O2` is usually a good tradeoff between size and performance, and `-O3` sometimes can even perform worse (see the entry for x265 [here](https://www.phoronix.com/review/11900k-gcc11-clang12/3)).

**Comment by albinahlback** [comment]
> The main problem is that documentation is scarce or cryptic (also an issue for parts of the mpoly code).

Yeah. We also need to think if we need functions like `n_min` in `fft_small.h`.

**Comment by fredrik-johansson** [comment]
> So I think we definitely should optimize this module, but we need to create a profiler in order to see what works. `-O2` is usually a good tradeoff between size and performance, and `-O3` sometimes can even perform worse (see the entry for x265 [here](https://www.phoronix.com/review/11900k-gcc11-clang12/3)).

I believe Dan already did some careful profiling to come up with the compiler flags he used, so it makes sense to use those flags as the starting point.

**Comment by fredrik-johansson** [comment]
> Yeah. We also need to think if we need functions like `n_min` in `fft_small.h`.

As helper functions internal to ``fft_small.h``, I think they are fine. I suppose we can move the ones that are useful elsewhere to ``ulong_extras.h``.

**Comment by fredrik-johansson** [comment]
@albinahlback The only thing missing to merge as far as I'm concerned is defining ``FLINT_HAVE_FFT_SMALL``.

It would be good to merge this as we can easily speed up a lot of FLINT 2x-3x with it.

We can open separate issues for:
* Cleaning up / documenting the code
* Enabling AVX2/Neon and fft_small by default on machines that support it
* Profiling / compiler flags

**Comment by albinahlback** [comment]
Like this?

**Comment by fredrik-johansson** [comment]
Yeah, that looks fine.

**Comment by fredrik-johansson** [comment]
Any idea why we get ``make: *** [Makefile:679: build/fq/test/t-is_primitive_TEST_RUN] Segmentation fault (core dumped)`` here on CI?

**Comment by fredrik-johansson** [comment]
Nevermind, that failure also occurs randomly in the trunk branch. Uninitialized fmpz in fq_is_primitive.

**Comment by fredrik-johansson** [comment]
Looks good to merge?

**Comment by albinahlback** [comment]
I suppose so. When/if you merge, can you open issues related to it? I suppose documentation is a priority.

**Comment by fredrik-johansson** [comment]
Sure.

**Comment by fredrik-johansson** [comment]
Thanks for getting this done!

**Pull request merged**